File: src/main/java/pt/haslab/dql/tictactoe/agents/BasicAgent.java
package pt.haslab.dql.tictactoe.agents;

import pt.haslab.dql.tictactoe.game.Actions;
import pt.haslab.dql.tictactoe.game.Board;
import pt.haslab.dql.tictactoe.game.Seed;

import java.util.Random;

/**
 * Implements an agent that attempts to play obvious and
 * typical moves if possible or random ones otherwise.
 * Created by nunomachado on 07/08/17.
 */
public class BasicAgent extends Agent {

    public BasicAgent(){}

    //agent acts with most common winning moves
    public String act(Board state) {

        boolean middleEmpty = state.cells[1][1].content == Seed.EMPTY;
        boolean middleNought = state.cells[1][1].content == Seed.NOUGHT;
        boolean topRightEmpty = state.cells[0][2].content == Seed.EMPTY;
        boolean topRightNought = state.cells[0][2].content == Seed.NOUGHT;
        boolean topLeftEmpty = state.cells[0][0].content == Seed.EMPTY;
        boolean topLeftNought = state.cells[0][0].content == Seed.NOUGHT;
        boolean bottomLeftEmpty = state.cells[2][0].content == Seed.EMPTY;
        boolean bottomRightEmpty = state.cells[2][2].content == Seed.EMPTY;

        //mark the middle first - "2 2"
        if(middleEmpty)
            return Actions.ACTIONS[4];
        //mark the top right - "1 3"
        if(middleNought && topRightEmpty)
            return Actions.ACTIONS[2];
        //mark the top right - "3 1"
        if(middleNought && topRightNought && bottomLeftEmpty)
            return Actions.ACTIONS[6];
        //mark the top left - "1 1"
        if(middleNought && topLeftEmpty)
            return Actions.ACTIONS[0];
        //mark the bottom right - "3 3"
        if(middleNought && topLeftNought && bottomRightEmpty)
            return Actions.ACTIONS[8];

        //play randomly otherwise
        Random r = new Random();
        int a = r.nextInt(Actions.ACTIONS.length);
        return Actions.ACTIONS[a];
    }
}

File: tictactoe/agents/basic_agent.py
import random
from tictactoe.agents.agent import Agent
from tictactoe.game.moves import Moves
from tictactoe.game.board import Board, Seed

class BasicAgent(Agent):

    def make_move(self, state: Board) -> str:
        middle_empty = state.cells[1][1].content == Seed.EMPTY
        middle_nought = state.cells[1][1].content == Seed.NOUGHT
        top_right_empty = state.cells[0][2].content == Seed.EMPTY
        top_right_nought = state.cells[0][2].content == Seed.NOUGHT
        top_left_empty = state.cells[0][0].content == Seed.EMPTY
        top_left_nought = state.cells[0][0].content == Seed.NOUGHT
        bottom_left_empty = state.cells[2][0].content == Seed.EMPTY
        bottom_right_empty = state.cells[2][2].content == Seed.EMPTY
        if middle_empty:
            return Moves.MOVES[4]
        if middle_nought and top_right_empty:
            return Moves.MOVES[2]
        if middle_nought and top_right_nought and bottom_left_empty:
            return Moves.MOVES[6]
        if middle_nought and top_left_empty:
            return Moves.MOVES[0]
        if middle_nought and top_left_nought and bottom_right_empty:
            return Moves.MOVES[8]
        return random.choice(Moves.MOVES)

    def __str__(self):
        return 'BASIC'